lib/eslint/eslint-helpers.js

```diff
@@ -277,94 +256,100 @@ async function globSearch({
-
-                    // entries may be directories or files so filter out directories
-                    if (entry.dirent.isDirectory()) {
-                        return false;
-                    }
+                // no match, so we can skip this directory
+                if (!matchesPattern) {
+                    return false;
+                }
 
-                    /*
-                     * Optimization: We need to track when patterns are left unmatched
-                     * and so we use `unmatchedPatterns` to do that. There is a bit of
-                     * complexity here because the same file can be matched by more than
-                     * one pattern. So, when we start, we actually need to test every
-                     * pattern against every file. Once we know there are no remaining
-                     * unmatched patterns, then we can switch to just looking for the
-                     * first matching pattern for improved speed.
-                     */
-                    const matchesPattern = unmatchedPatterns.size > 0
-                        ? matchers.reduce((previousValue, matcher) => {
-                            const pathMatches = matcher.match(relativePath);
-
-                            /*
-                             * We updated the unmatched patterns set only if the path
-                             * matches and the file has a config. If the file has no
-                             * config, that means there wasn't a match for the
-                             * pattern so it should not be removed.
-                             *
-                             * Performance note: `getConfig()` aggressively caches
-                             * results so there is no performance penalty for calling
-                             * it multiple times with the same argument.
-                             */
-                            if (pathMatches && configs.getConfig(entry.path)) {
-                                unmatchedPatterns.delete(matcher.pattern);
-                            }
-
-                            return pathMatches || previousValue;
-                        }, false)
-                        : matchers.some(matcher => matcher.match(relativePath));
-
-                    return matchesPattern && configs.getConfig(entry.path) !== void 0;
-                })
+                const absolutePath = normalizeToPosix(path.resolve(basePath, entry.path));
+
+                /*
+                 * Special case: Don't look for a config file when "." is passed. This
+                 * is because we don't want to ignore the directory itself, just its
+                 * contents.
+                 */
+                const configs = rawPatterns[matchIndex] !== "."
+                    ? void 0
+                    : await configLoader.loadConfigArrayForDirectory(absolutePath);
+
+                return !configs?.isDirectoryIgnored(absolutePath);
```

Comment:
Now, on this branch, a pattern like `**/*.js` would traverse all directories, which is not what we want, I believe.

To reproduce, add `console.log(entry.path);` at the beginning of `async directoryFilter(entry)` and run `npx eslint "**/*.js"`.

```
.cache
.git
.git/hooks
.git/info
.git/logs
.git/logs/refs
.git/logs/refs/heads
.git/logs/refs/remotes
.git/logs/refs/remotes/origin
.git/logs/refs/remotes/origin/chore
.git/logs/refs/remotes/origin/docs
.git/logs/refs/remotes/origin/fix
.git/logs/refs/remotes/origin/renovate
.git/objects
.git/objects/00
.git/objects/01
.git/objects/02
.git/objects/03
.git/objects/04
.git/objects/05
.git/objects/06
.git/objects/07
.git/objects/08
.git/objects/09
.git/objects/0a
.git/objects/0b
.git/objects/0c
.git/objects/0d
.git/objects/0e
.git/objects/0f
.git/objects/10
.git/objects/11
.git/objects/12
.git/objects/13
.git/objects/14
.git/objects/15
.git/objects/16
.git/objects/17
.git/objects/18
.git/objects/19
.git/objects/1a
.git/objects/1b
.git/objects/1c
.git/objects/1d
.git/objects/1e
.git/objects/1f
.git/objects/20
.git/objects/21
.git/objects/22
.git/objects/23
.git/objects/24
.git/objects/25
.git/objects/26
.git/objects/27
.git/objects/28
.git/objects/29
.git/objects/2a
.git/objects/2b
.git/objects/2c
.git/objects/2d
.git/objects/2e
.git/objects/2f
.git/objects/30
.git/objects/31
.git/objects/32
.git/objects/33
.git/objects/34
.git/objects/35
.git/objects/36
.git/objects/37
.git/objects/38
.git/objects/39
.git/objects/3a
.git/objects/3b
.git/objects/3c
.git/objects/3d
.git/objects/3e
.git/objects/3f
.git/objects/40
.git/objects/41
.git/objects/42
.git/objects/43
.git/objects/44
.git/objects/45
.git/objects/46
.git/objects/47
.git/objects/48
.git/objects/49
.git/objects/4a
.git/objects/4b
.git/objects/4c
.git/objects/4d
.git/objects/4e
.git/objects/4f
.git/objects/50
.git/objects/51

... (~10,000 directories)
```

Reply:
Hmmm yeah, that's not right. I'll take a look.

Reply:
I'm not able to reproduce this (Windows 10). Are you sure you're using the right command (`npx eslint` would download ESLint if you don't have it locally?). Here's what I'm getting:

```
$ node ./bin/eslint.js "**/*.js"
(node:5064) ESLintIgnoreWarning: The ".eslintignore" file is no longer supported. 
Switch to using the "ignores" property in "eslint.config.js": https://eslint.org/docs/latest/use/configure/migration-guide#ignoring-files
(Use `node --trace-warnings ...` to show 
where the warning was created)
.cache
.git
.github
.nyc_output
.trunk
.vscode
bin
build
conf
coverage
docs
lib
messages
node_modules
packages
templates
tests
tmp
tools

Oops! Something went wrong! :(

ESLint: 9.9.1

No files matching the pattern "**C:/Program Files/Git/*.js" were found.
Please check for typing mistakes in the pattern.
```

Reply:
> Oops! Something went wrong! :(
> 
> ESLint: 9.9.1
> 
> No files matching the pattern "**C:/Program Files/Git/*.js" were found.
> Please check for typing mistakes in the pattern.

That `"**C:/Program Files/Git/*.js"` looks weird, I'm not sure where it did come from when running `$ node ./bin/eslint.js "**/*.js"` (maybe your shell is doing some expansion despite the quotes?).

I tried again, on this branch, with a clean `npm install`, and in cmd.exe which doesn't expand globs. I also added `console.log({ patterns });` at the beginning of `async function globSearch` to make sure that the pattern is passed through.

```js
C:\projects\eslint>node ./bin/eslint.js "**/*.js"
(node:7136) ESLintIgnoreWarning: The ".eslintignore" file is no longer supported. Switch to using the "ignores" property in "eslint.config.js": https://eslint.org/docs/latest/use/configure/migration-guide#ignoring-files
(Use `node --trace-warnings ...` to show where the warning was created)

{ patterns: [ 'C:\\projects\\eslint\\**\\*.js' ] }
.cache
.git
.git/hooks
.git/info
.git/logs
.git/logs/refs
.git/logs/refs/heads
.git/logs/refs/remotes
.git/logs/refs/remotes/origin
.git/logs/refs/remotes/origin/chore
.git/logs/refs/remotes/origin/docs
.git/logs/refs/remotes/origin/fix
.git/logs/refs/remotes/origin/renovate
.git/objects
.git/objects/00
.git/objects/01
.git/objects/02
.git/objects/03
.git/objects/04
.git/objects/05
.git/objects/06
.git/objects/07
.git/objects/08
.git/objects/09
.git/objects/0a
.git/objects/0b
.git/objects/0c

...
```

Reply:
Okay, I can see this in `cmd` but not in Git Bash. It looks like Git Bash is representing the path in Windows differently:

```
  patterns: [
    'C:\\Users\\nzaka\\projects\\eslint\\eslint\\**C:\\Program Files\\Git\\*.js'  
  ]
```

I'll take a look at that, too.

Reply:
So it looks like Git for Windows is expanding the pattern even if it's quoted.
https://github.com/git-for-windows/git/issues/5131